WIP: Validate metadata for custom resources

nikhita opened the pull request:
Validates the following by converting the metadata subtree of custom resources from `Unstructured` to a metadata struct and back:

1. null/nil values for labels and annotations are marshalled as empty strings `""`.
2. Validates metadata to make sure there are no unknown fields.
3. Validates invalid values (numbers, booleans, structured values, etc) for labels, annotations, etc.

Fixes #59451

TODO:

- [ ] Decide how to proceed with existing bad data

**Release note**:

(This probably deserves a release note but will update after the TODO is done)
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

/cc sttts liggitt deads2k ash2k jennybuckley

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/validation_test.go:
```diff
@@ -30,61 +30,6 @@ import (
 	"k8s.io/apiextensions-apiserver/test/integration/testserver"
 )
 
-func TestForProperValidationErrors(t *testing.T) {
```
This is a pure move.

liggitt commented:
Is there a way to use the `unstructured<->typed` converter to convert the metadata subtree to a real Metadata struct, then back, in PrepareForCreate and PrepareForUpdate?

wgliang commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go:
```diff
@@ -0,0 +1,146 @@
+	objectMetaFieldNames := make(map[string]bool, totalFields)
+
+	for i := 0; i < totalFields; i++ {
+		jsonName := strings.Split(v.Type().Field(i).Tag.Get("json"), ",")[0]
```
This index approach should be added to determine the safety of the approach, or there may be a risk of panic.

wgliang commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go:
```diff
@@ -0,0 +1,146 @@
+	}
+
+	customResourceObject, ok := obj.(*unstructured.Unstructured)
+	// this will never happen.
```
I suggest to remove this comment.

sttts replied:
or better move it inside the if clause.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go:
```diff
@@ -0,0 +1,146 @@
+func ValidateObjectMetaFields(objectMeta map[string]interface{}) error {
+	emptyObjectMeta := metav1.ObjectMeta{}
+
+	// detect unknown fields in metadata
```
this isn't the same behavior as native resources... we should drop unknown fields, not error on them (otherwise our API server is not forward compatible with ignoring new metadata fields)

ash2k commented:
> Is there a way to use the unstructured<->typed converter to convert the metadata subtree to a real Metadata struct, then back, in PrepareForCreate and PrepareForUpdate?

I think that would be the best approach. Can be done using `k8s.io/apimachinery/pkg/runtime/DefaultUnstructuredConverter.(From|To)Unstructured()`.

nikhita committed:
Validate metadata for custom resources

Metadata subtree for custom resources is converted from
Unstructured i.e. map[string]interface{} to a metadata struct
and back for the following validations:

- null/nil values for labels and annotations are marshalled as empty strings "".
- validate metadata to make sure there are no unknown fields.
- validate invalid values (numbers, booleans, structured values, etc)
for labels, annotations, etc.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
+	// if there are errors, we don't convert and fail in ValidateCreate and ValidateUpdate
+	in := metadata.(map[string]interface{})
+	out := &metav1.ObjectMeta{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(in, out); err == nil {
```
return early rather than nesting

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
 
 func (customResourceDefinitionStorageStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, old runtime.Object) {
+	customResourceObject, _ := obj.(*unstructured.Unstructured)
+	metadata := customResourceObject.UnstructuredContent()["metadata"]
```
`, ok` and return early if missing

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -115,73 +136,3 @@ func (a customResourceDefinitionStorageStrategy) MatchCustomResourceDefinitionSt
 	}
 }
-
-type customResourceValidator struct {
```
hard to follow changes combined with the move... can you keep it in place or separate the straight move into a separate commit?

nikhita replied:
can keep it in place. the move is in the subresources PR (adds lots of things there).

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:
```diff
@@ -63,9 +58,35 @@ func (a customResourceDefinitionStorageStrategy) NamespaceScoped() bool {
+	// convert metadata subtree to metadata struct and back
+	// this is done to make sure we drop unknown fields from metadata
+	// if there are errors, we don't convert and fail in ValidateCreate and ValidateUpdate
+	in := metadata.(map[string]interface{})
```
`, ok` and short-circuit... hard casting here can panic

liggitt replied:
probably want a helper that does this coercion so we can call it from PrepareForCreate, PrepareForUpdate, and validation functions

that would allow unit testing various scenarios as well

I had a WIP doing something similar in https://github.com/liggitt/kubernetes/commits/crd-coercion in case it's useful.

liggitt replied:
also, based on the issues I was hitting, I don't think fromUnstructured does the nil -> "" conversion we wanted

nikhita replied:
`structuredMetadata` for the `ValidateMetadata` test:

```
&ObjectMeta{Name:foo,GenerateName:,Namespace:not-the-default,SelfLink:/apis/mygroup.example.com/v1beta1/namespaces/not-the-default/noxus/foo,UID:169ff77e-158a-11e8-9dd2-54e1ad6c2d05,ResourceVersion:6,Generation:0,CreationTimestamp:2018-02-19 21:02:26 +0530 IST,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{qux: ,},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,}
```

nikhita replied:
> also, based on the issues I was hitting, I don't think fromUnstructured does the nil -> "" conversion we "wanted"

Also, from this test data - https://github.com/liggitt/kubernetes/commit/51965f9e381e9588d918ad1281bf94b1bec65c58#diff-1b3f35816cc25f4fd5cd83f1c18c013fR327:

The unstructured object is:
```
map[bb:<nil> bc:<nil> bd:<nil> ba:map[aa:<nil> ab:<nil> ac:<nil>]]
```

and both `newObj` and `unmarshalledObj` are:

```
{"ba":{},"bb":"","bc":null,"bd":null}
```

which seems right to me. @liggitt thoughts?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go:
```diff
@@ -0,0 +1,130 @@
+	return validation.ValidateObjectMetaAccessor(accessor, a.namespaceScoped, validation.NameIsDNSSubdomain, field.NewPath("metadata"))
+}
+
+func (a customResourceValidator) ValidateUpdate(ctx genericapirequest.Context, obj, old runtime.Object) field.ErrorList {
```
need to think about what we do here, since we have the possibility of persisted mis-structured data.

ratcheting validation seems like a good start (if the old object coerces without error, require the new object to as well).

dropping fields with structural errors (along with unknown fields) is another possibility (it's likely what we would try to do to fix the issues), but is not as simple to do.

nikhita replied:
this is still open.

nikhita commented:
/sig api-machinery
/area custom-resources

nikhita commented:
Closing in favor of pruning: https://github.com/kubernetes/kubernetes/pull/63604.

Also, the pruning PR exists behind a feature gate so it takes care of the following TODO:

> Decide how to proceed with existing bad data

/close
